designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+    * For e.g., external plugin `helm.sdk.operatorframework.io/v2-alpha` can be specified through the `--plugins` flag for every subcommand and also be defined in the project configuration in the `layout` field for plugin resolution.
+
+Example `PROJECT` file:
+
+```yaml
+version: "3"
+domain: testproject.org
+layout: 
+- go.kubebuilder.io/v3
+- helm.sdk.operatorframework.io/v2-alpha
+plugins:
+  helm.sdk.operatorframework.io/v2-alpha:
+    resources:
+    - domain: testproject.org
+      group: crew
+      kind: Captain
+      version: v2-alpha
+  declarative.go.kubebuilder.io/v1:
+    resources:
+    - domain: testproject.org
+      group: crew
+      kind: FirstMate
+      version: v1
+repo: github.com/test-inc/testproject
+resources:
+- group: crew
+  kind: Captain
+  version: v1
+```
+
+#### What to pass between Kubebuilder and plugins
+
+<!--- TODO(rashmigottipati): This section needs more details on what structure to pass to the plugins. --->
+
+Currently, the `injector` type is defined as below:
+
+```go
+// injector is used to inject certain fields to file templates.
+type injector struct {
+    // config stores the project configuration.
+    config config.Config
+
+    // boilerplate is the copyright comment added at the top of scaffolded files.
+    boilerplate string
+
+    // resource contains the information of the API that is being scaffolded.
+    resource *resource.Resource
+}
+```
+
+One of the plugin hooks allows plugins to set up their own flags. Commong flags (like `--group`, `--version`, `--kind`) are defined by the plugin system itself, and the resource created with those values populated will be passed to the plugins. `create api` and `create webhook` commands would require the GVK being passed as well to the plugins. Currently, `InjectResource` is a required hook for `create api` and `create webhook` commands and the entire resource (which contains the group, version and kind) is being passed to the plugins for these commands.
+
+All the raw flags of the plugins would have to be passed down to the plugins through injector. The plugin specific flags would have to be specified in the project configuration by the user. And the current implmentation is such that, every subcommand can optionally implement the `InjectConfig` hook which passes the project configuration to the plugins and also enable the plugins to modify it if required.
+
+Along with flags, boilerplate with empty copyright or an apache license and in go-style comments is also being passed down to the plugins through injector. If other kind of comments were to be supported, this logic would have to be added (scaffold markers do support YAML and Go-style comments).
+
+Additionally, we would pass the entire CLI environment to the plugin, as anything executing in subshell environment would be able to read the environment variables even if it's being restricted in terms of what Kubebuilder passes to it.
```

Comment:
IIRC I said this is not true, and if I didn't I apologize. The child process' environment is completely controlled by the parent process. Regardless I'd say for simplicity we pass the environment to the child process. Of course this brings up security concerns, but you already agree to execute arbitrary code by specifying an external plugin so this is not a top priority imo.
```suggestion
Additionally, we would pass the entire CLI environment to the plugin.
```